src/input/source/led/multicolor.rs

```diff
@@ -0,0 +1,272 @@
+
+    #[error("Error writing value: {0}")]
+    Write(std::io::Error),
+
+    #[error("Path not found: {0}")]
+    Path(PathBuf),
+
+    #[error("Unsupported index type: {0}")]
+    UnsupportedIndex(String),
+}
+
+enum IndexType {
+    #[allow(clippy::upper_case_acronyms)]
+    RGB,
+    Red,
+    Green,
+    Blue,
+}
+
+/// MultiColorChassis source device implementation
+pub struct LedMultiColor {
+    brightness_path: PathBuf,
+    current_brightness: u8,
+    current_color: LedFixedColor,
+    max_brightness: u8,
+    multi_index_map: Vec<IndexType>,
+    multi_intensity_path: PathBuf,
+}
+
+impl LedMultiColor {
+    /// Create a new LedMultiColor source device with the given udev
+    /// device information
+    pub fn new(
+        device_info: UdevDevice,
+        fixed_color: Option<LedFixedColor>,
+    ) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        let brightness_path = PathBuf::from(device_info.syspath().as_str()).join("brightness");
+        if !brightness_path.exists() {
+            return Err(Box::new(LedMcError::Path(brightness_path)));
+        }
+
+        let current_brightness: u8 = read_brightness(brightness_path.as_path())?;
+
+        let max_brightness_path =
+            PathBuf::from(device_info.syspath().as_str()).join("max_brightness");
+        if !max_brightness_path.exists() {
+            return Err(Box::new(LedMcError::Path(brightness_path)));
+        }
+
+        let max_brightness: u8 = read_brightness(max_brightness_path.as_path())?;
+
+        let multi_intensity_path =
+            PathBuf::from(device_info.syspath().as_str()).join("multi_intensity");
+        if !multi_intensity_path.exists() {
+            return Err(Box::new(LedMcError::Path(multi_intensity_path)));
+        }
+
+        let multi_index_path = PathBuf::from(device_info.syspath().as_str()).join("multi_index");
+        let multi_index_map = read_color_index(&multi_index_path)?;
+
+        let current_color = match fixed_color {
+            Some(color) => color,
+            None => read_color(multi_intensity_path.as_path(), &multi_index_map)?,
+        };
+
+        log::debug!(
+            "LED MultiColor Device setup complete. brightness: {}, max_brightness {}, color {:?},",
+            current_brightness,
+            max_brightness,
+            current_color
+        );
+
+        let result = Self {
+            brightness_path,
+            current_brightness,
+            current_color,
+            max_brightness,
+            multi_index_map,
+            multi_intensity_path,
+        };
+
+        // Immediatly set the user-defined color if one is provided
+        result.write_color(
+            result.current_color.r,
+            result.current_color.g,
+            result.current_color.b,
+        )?;
+
+        Ok(result)
+    }
+
+    /// Writes the given red, green, and blue values to the device.
+    fn write_color(&self, r: u8, g: u8, b: u8) -> Result<(), Box<dyn Error + Send + Sync>> {
```

Comment:
To support non-RGB multi-color LEDs, I think we should change this method to take a hashmap of values instead of hard-coding RGB values. E.g.

```suggestion
    fn write_color(&self, color: HashMap<String, u8>) -> Result<(), Box<dyn Error + Send + Sync>> {
```

Then write the values according to the `multi_index` mapping.

Reply:
This function is still useful for target interfaces that only support RGB writing. A more comprehensive write_colors function was added in v2.